arwen/contexts/guard.go

```diff
@@ -0,0 +1,69 @@
+package contexts
+
+import (
+	"fmt"
+	"unicode"
+
+	"github.com/ElrondNetwork/arwen-wasm-vm/arwen"
+	"github.com/ElrondNetwork/arwen-wasm-vm/wasmer"
+)
+
+func (context *runtimeContext) VerifyContractCode() error {
+	err := verifyMemoryDeclaration(context.instance)
+	if err != nil {
+		return err
+	}
+
+	err = verifyFunctionsNames(context.instance)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func verifyMemoryDeclaration(instance *wasmer.Instance) error {
+	if !instance.HasMemory() {
+		return arwen.ErrMemoryDeclarationMissing
+	}
+
+	return nil
+}
+
+func verifyFunctionsNames(instance *wasmer.Instance) error {
+	for functionName := range instance.Exports {
+		if !isValidFunctionName(functionName) {
+			return fmt.Errorf("%v: %s", arwen.ErrInvalidFunctionName, functionName)
+		}
+	}
+
+	return nil
+}
+
+func isValidFunctionName(functionName string) bool {
+	const maxLengthOfFunctionName = 256
+
+	if len(functionName) == 0 {
+		return false
+	}
+
+	if len(functionName) >= maxLengthOfFunctionName {
+		return false
+	}
+
```

Comment:
delete empty line

Reply:
we might add a few functions will be unable to be called, builtin elrond functions. I would add in arwen constructor a map[string]struct{} which function names are taken. One example right now is "claimDeveloperRewards"

Reply:
I've added a new struct called `ReservedFunctions`. If this is ok, I can move it to `vm-common`, so that the protocol can also reference it in a clean manner.